src/kudu/consensus/consensus_queue.cc

```diff
@@ -164,13 +166,15 @@ PeerMessageQueue::Metrics::Metrics(const scoped_refptr<MetricEntity>& metric_ent
 
 PeerMessageQueue::PeerMessageQueue(const scoped_refptr<MetricEntity>& metric_entity,
                                    scoped_refptr<log::Log> log,
+                                   scoped_refptr<ConsensusMetadata> cmeta,
```

Comment:
which cmeta is this. cmeta can change as new instances are added and removed to the ring.

Reply:
We should not need `cmeta` to get the data commit quorum and it's not really safe to do so because it's protected by `RaftConsensus::lock_` (see header file comment indicating it's not thread-safe). We get the latest active `RaftConfigPB` every time `PeerMessageQueue::SetLeaderMode()` is called, which is called any time we should care (on leader change and raft config change). If we can, we should simply include any metadata needed to make commitment decisions in `RaftConfigPB`, which is the purpose of that struct.

If that's not sufficient, due to the data commit quorum changing on leader change without raft config change, we should try to plumb a new struct into `SetLeaderMode()` for this purpose and figure out how we are going to persist the relevant metadata (most likely in cmeta but not directly accessed from this class).

Reply:
I think you guys are right here. I went with it because I also needed the term for when the last time quorums were resolved in order to ensure safety for severely lagging followers. Let me remove this from the data commit quorum diff and address it in the diff that I am working on and ties the config pieces together so that there is less confusion but like you mentioned it might likely end up in ConsensusMetadataPB but there might be safer options. Let me take a second look if I can simplify that diff further.